dali/pipeline/executor/executor2/op_task.cc

```diff
@@ -0,0 +1,366 @@
+  for (int i = 0; i < ws_->NumInput(); i++, ti++) {
+    if (ws_->InputIsType<CPUBackend>(i)) {
+      auto inp = TaskInput<CPUBackend>(ti);
+      if (inp.event && inp.order != order)
+        events.insert(inp.event);
+      ws_->SetInput(i, inp.data);
+    } else {
+      assert(ws_->InputIsType<GPUBackend>(i));
+      auto inp = TaskInput<GPUBackend>(ti);
+      // If the output order of the operator is `host` then we don't wait for GPU
+      // inputs - they can't be accessed directly on host and the operator will
+      // have to issue some sort of synchronization if and when necessary.
+      // This optimization is essential to avoid oversynchronization
+      // when the operator needs to access the metadata only (e.g. getting the shape).
+      if (order.is_device() /*see comment above */ && inp.event && inp.order != order)
+        events.insert(inp.event);
+      ws_->SetInput(i, inp.data);
+    }
+  }
+
+  for (int i = 0; i < ws_->NumArgumentInput(); i++, ti++) {
+    auto &inp = TaskInput<CPUBackend>(ti);
+    if (inp.event)
+      events.insert(inp.event);
+    ws_->SetArgumentInput(i, inp.data);
+  }
+
+  if (task_->NumInputs() > ti) {  // Artificial error input
+    // TODO(michalz): Add early failure mechanism to `dali::tasking` that makes the successors
+    //                of a task that threw an exception fail. Currently the error disappears,
+    //                so we have to add an artificial input which propagates the exception.
+    task_->GetInputValue<void>(ti++);
+  }
+
+  for (auto e : events)
+    ws_->output_order().wait(e);
+
+  if (ws_->NumOutput()) {
+    if (ws_->NumInput() > 0) {
+      ws_->SetBatchSizes(ws_->GetInputBatchSize(0));
+    } else if (ws_->NumArgumentInput() > 0) {
+      ws_->SetBatchSizes(ws_->ArgumentInput(0).num_samples());
+    } else {
+      ws_->SetBatchSizes(ws_->GetIterationData()->default_batch_size);
+    }
+  }
+}
+
+AccessOrder OpTask::OutputConsumerOrder(int output_idx) {
+  assert(static_cast<size_t>(output_idx) < node_->outputs.size());
+  // Return the common stream.
+  auto &consumers = node_->outputs[output_idx].consumers;
+  if (consumers.empty())
+    return {};  // definitely no consumer
+  AccessOrder order = consumers[0]->consumer->env.order;
+  for (size_t i = 1; i < consumers.size(); i++)
+    if (consumers[i]->consumer->env.order != order)
+      return {};
+  return order;
+}
+
+OpTask::OpTaskOutputs OpTask::GetWorkspaceOutputs() {
+  OpTaskOutputs ret;
+  int nout = ws_->NumOutput();
+  ret.reserve(nout);
+  cudaEvent_t event = ws_->has_event() ? ws_->event() : nullptr;
+  auto order = ws_->output_order();
+  for (int o = 0; o < nout; o++) {
+    if (ws_->OutputIsType<CPUBackend>(o)) {
+      auto ptr = ws_->OutputPtr<CPUBackend>(o);
+      if (!ptr->shares_data()) {
+        if (AccessOrder consumer_order = OutputConsumerOrder(o))
+          ptr->set_order(consumer_order, false);
```

Comment:
What happens on backend boundaries? 
What if we have for example CPU order (or some Mixed-based order) and all consumers are GPU-based, what does setting the order without sync mean in that case?
Why can we do it only where there is no data sharing?
We won't be writing to those buffers again in this executor, right? So it's just matter of consuming them after the producer finishes?

Reply:
I need to think about this one. So far it works (and L0 passes with this executor as the default), but it looks a bit suspicious...

Reply:
OK, more on that:
The function `OutputConsumerOrder` gets the unique consumption order, if such an order exists - or returns null order otherwise.
When all consumers are in the same order/stream, we can just move the object to that order, so that it's deleted in that order. This simplifies the lifecycle.
If there's no unique order, the production order and consumption order remain different. The synchronization is inserted in `OpTask::SetWorkspaceInputs`, but the deletion semantics are enforced by waiting for consumption in the producer stream. This guarantees that when the buffer is deleted in the production stream, the deletion will be sequenced after consumption.

Reply:
Thanks. Btw, maybe put this here as a comment?

Reply:
Done.